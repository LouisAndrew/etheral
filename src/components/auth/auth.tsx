import React, { useState, useEffect } from 'react';
import { useNavigate } from '@reach/router';
import { useDispatch } from 'react-redux';

import { Flex, Box, Button } from 'rebass';

import Login from './login';
import SignUp from './signup';
import useAllProducts from 'helper/use-all-products';
import { FirebaseUserData, InCart } from 'helper/schema/firebase-user';
import { setCart } from 'state/actions/cart';
import extractCartFirestore from 'helper/extract-cart-firestore';

type Props = {
    auth: firebase.auth.Auth;
    db: firebase.firestore.Firestore;
    googleProvider: firebase.auth.GoogleAuthProvider;
    authPersistence: string;
};

export type LoginProps = {
    email: string;
    password: string;
};

export type SignUpProps = LoginProps & {
    name: string;
    signupNewsletter: boolean;
};

// enum to show where the error occurs.
enum ErrorTarget {
    NONE,
    LOGIN,
    SIGN_UP,
}

const Auth: React.FC<Props> = ({
    auth,
    db,
    googleProvider,
    authPersistence,
}) => {
    const [uid, setUid] = useState('');
    const [isNewUser, setIsNewUser] = useState(false);

    // state to store errror msg.
    const [error, setError] = useState<firebase.auth.Error | undefined>(
        undefined
    );
    const [errTarget, setErrTarget] = useState<ErrorTarget>(ErrorTarget.NONE);

    const navigate = useNavigate();
    const allProducts = useAllProducts();
    const dispatch = useDispatch();

    // set auth persistence.
    useEffect(() => {
        auth.setPersistence(authPersistence);
    }, []);

    useEffect(() => {
        // call this function just if user is logged in!
        const currentUser = auth.currentUser;
        if (currentUser !== null) {
            (async () => {
                if (!uid) {
                    await setUid(currentUser.uid);
                }

                if (!isNewUser) {
                    await fetchCartItems(currentUser);
                } else {
                    await createNewUser(currentUser);
                }

                await navigate('/');
            })();
        }
    }, [uid]);

    const createNewUser = async (currentUser: firebase.User) => {
        if (currentUser !== null) {
            try {
                const userData: FirebaseUserData = {
                    name: currentUser.displayName
                        ? currentUser.displayName
                        : '',
                    email: currentUser.email ? currentUser.email : '',
                    inCart: [],
                    wishlist: [],
                    orders: [],
                };

                await db
                    .collection('user')
                    .doc(currentUser.uid)
                    .set(userData);
            } catch (e) {
                console.error(e);
            }
        }
    };

    const fetchCartItems = async (currentUser: firebase.User) => {
        // fetch items on cart and wishlist of the user => on login.
        if (currentUser) {
            try {
                const docRef = await db
                    .collection('user')
                    .doc(currentUser.uid)
                    .get();

                const userData = await docRef.data();

                // fetch and checks for all undefined object.
                if ((await docRef.exists) && userData) {
                    const inCart = (await (userData.inCart as any)) as InCart;
                    const onWishlist = (await (userData.wishlist as any)) as InCart;

                    const filteredInCartData = await extractCartFirestore({
                        firestoreCartData: inCart,
                        allProducts,
                    });
                    const filteredWishlistData = await extractCartFirestore({
                        firestoreCartData: onWishlist,
                        allProducts,
                    });

                    await dispatch(
                        setCart({
                            cart: filteredInCartData,
                            wishlist: filteredWishlistData,
                        })
                    );
                }
            } catch (e) {
                console.error(e);
            }
        }
    };

    // user login
    const loginWithEmail = async ({ password, email }: LoginProps) => {
        try {
            const loginReq = await auth.signInWithEmailAndPassword(
                email,
                password
            );
            const { user } = await loginReq;

            if (user !== null) {
                setUid(user.uid);
            }
        } catch (e) {
            setErrTarget(ErrorTarget.LOGIN);
            setError(e);
        }
    };

    // create new user
    const signupWithEmail = async ({
        name,
        password,
        email,
        signupNewsletter,
    }: SignUpProps) => {
        try {
            const signupReq = await auth.createUserWithEmailAndPassword(
                email,
                password
            );

            const { user } = await signupReq;

            if (user !== null) {
                await user.updateProfile({
                    displayName: name,
                });
                await setIsNewUser(true);
                await setUid(user.uid);
            }

            if (await signupNewsletter) {
                // call cloud function to sign this acc to newsletter up
                console.log('Do signup newsletter here');
            }
        } catch (e) {
            setErrTarget(ErrorTarget.SIGN_UP);
            setError(e);
        }
    };

    // sign in/up with google
    const withGoogle = async () => {
        try {
            const signinReq = await auth.signInWithPopup(googleProvider);

            const { user } = await signinReq;

            if (user !== null) {
                if (signinReq.additionalUserInfo) {
                    await setIsNewUser(signinReq.additionalUserInfo.isNewUser);
                }
                await setUid(user.uid);
            }
        } catch (e) {
            console.error(e);
        }
    };

    return (
        <Flex
            width={['100%', '100%', '100%', '70%', '60%']}
            justifyContent="space-between"
            flexWrap="wrap"
            px={[4, 4, 6]}
            m="0 auto"
        >
            <Box width={['100%', '100%', '48%', '45%', '44%']} my={[5]}>
                <SignUp
                    signup={signupWithEmail}
                    firebaseError={
                        errTarget === ErrorTarget.SIGN_UP ? error : undefined
                    }
                />
            </Box>
            <Box
                width={['100%', '100%', '48%', '45%', '44%']}
                my={[5]}
                sx={{ textAlign: 'center' }}
            >
                <Login
                    login={loginWithEmail}
                    withGoogle={withGoogle}
                    firebaseError={
                        errTarget === ErrorTarget.LOGIN ? error : undefined
                    }
                />
            </Box>
        </Flex>
    );
};

export { Auth };
