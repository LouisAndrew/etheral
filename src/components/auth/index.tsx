import React from 'react';
import firebase from 'gatsby-plugin-firebase';

import { Auth as AuthEl } from './auth';

const Auth = <AuthEl auth={firebase.auth()} db={firebase.firestore()} />;

export default Auth;
