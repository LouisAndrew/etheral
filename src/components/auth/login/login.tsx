import React from 'react';
import { useForm } from 'react-hook-form';
import { debounce } from 'lodash';

import { Box, Heading } from 'rebass';
import { Input } from '@rebass/forms';

import { LoginProps } from '../auth';

type Props = {
    submitValue?: string;
    login: (args: LoginProps) => void;
};

type Inputs = {
    loginEmail: string;
    loginPassword: string;
};

const Login: React.FC<Props> = ({ login, submitValue }) => {
    const { register, handleSubmit } = useForm();

    const debouncedLogin = debounce(login, 500);

    const submit = ({ loginEmail, loginPassword }: Inputs) => {
        debouncedLogin({
            email: loginEmail,
            password: loginPassword,
        });
    };

    return (
        // <form onSubmit={handleSubmit(submit)}>
        //     <label htmlFor="login-email">Email</label>
        //     <input
        //         name="loginEmail"
        //         id="login-email"
        //         type="email"
        //         ref={register}
        //     />
        //     <label htmlFor="login-password">Password</label>
        //     <input
        //         name="loginPassword"
        //         id="login-password"
        //         type="password"
        //         ref={register}
        //     />
        //     <input type="submit" value={submitValue ? submitValue : 'Login'} />
        // </form>
        <Box as="form">
            <Heading as="h3" variant="formHeading">
                LOGIN
            </Heading>
        </Box>
    );
};

export { Login };
