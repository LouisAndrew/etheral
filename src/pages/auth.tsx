import React from 'react';
import { PageProps } from 'gatsby';
import { Layout } from 'components/layout';
import Auth from 'components/auth';

const AuthPage = (props: PageProps) => {
    return (
        <Layout>
            <div className="top">
                <Auth />
            </div>
        </Layout>
    );
};

export default AuthPage;
