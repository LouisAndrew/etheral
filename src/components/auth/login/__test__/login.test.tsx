import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { set } from 'lodash';
import userEvent from '@testing-library/user-event';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';

import { Login } from '../login';
import { LoginProps } from 'components/auth/auth';

describe('Login component', () => {
    const loginDataContainer = {
        email: '',
        password: '',
    };
    const login = jest.fn(({ email, password }: LoginProps) => {
        set(loginDataContainer, 'email', email);
        set(loginDataContainer, 'password', password);
    });

    const el: React.ReactElement = <Login login={login} />;

    afterEach(cleanup);

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(el, div);
    });

    it('should render and submits correctly', () => {
        const { getByLabelText, getByDisplayValue } = render(el);

        const emailInput: HTMLElement = getByLabelText('Email');
        const passwordInput: HTMLElement = getByLabelText('Password');

        const email = 'aa@aa';
        const password = '124';

        userEvent.type(emailInput, email);
        userEvent.type(passwordInput, password);

        const submitButton: HTMLElement = getByDisplayValue('Login');
        userEvent.click(submitButton);

        setTimeout(() => {
            expect(loginDataContainer.email).toBe(email);
            expect(loginDataContainer.password).toBe(password);
        }, 800);
    });

    it('matches snapshot', () => {
        const tree = renderer.create(el).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
