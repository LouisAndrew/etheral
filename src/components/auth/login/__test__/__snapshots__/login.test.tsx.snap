// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component matches snapshot 1`] = `
<form
  className="css-1lx7oj"
  onSubmit={[Function]}
>
  <h3
    className="css-d2e41n"
  >
    LOGIN
  </h3>
  <div
    className="css-vurnku"
  >
    <label
      className="css-knfhfe"
      htmlFor="login-email"
    >
      EMAIL
    </label>
    <input
      className="css-4y926w"
      id="login-email"
      name="loginEmail"
      type="email"
    />
    <label
      className="css-knfhfe"
      htmlFor="login-password"
    >
      PASSWORD
    </label>
    <input
      className="css-4y926w"
      id="login-password"
      name="loginPassword"
      type="password"
    />
  </div>
  <div
    className="css-b4n4pg"
    onClick={[Function]}
    role="button"
  >
    FORGOT PASSWORD?
  </div>
  <div
    className="css-lhwra4"
  >
    <button
      className="css-zypngu"
      type="submit"
    >
      LOGIN
    </button>
  </div>
</form>
`;
